private/tools/java/rules/jvm/external/maven/Outdated.java

```diff
@@ -0,0 +1,133 @@
+    String releaseVersion = null;
+
+    String url =
+        String.format("%s/%s/%s/maven-metadata.xml",
+            repository,
+            groupId.replaceAll("\\.", "/"),
+            artifactId);
+
+    DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
+    DocumentBuilder documentBuilder;
+    try {
+      documentBuilder = documentBuilderFactory.newDocumentBuilder();
+    } catch (ParserConfigurationException e) {
+      verboseLog(String.format("Caught exception %s", e));
+      return null;
+    }
+    Document document;
+    try {
+      document = documentBuilder.parse(new URL(url).openStream());
+    } catch (IOException | SAXException e) {
+      verboseLog(String.format("Caught exception for %s: %s", url, e));
+      return null;
+    }
+
+    // example maven-metadata.xml
+    //<metadata>
+    //  <versioning>
+    //    <latest>1.14.0-SNAPSHOT</latest>
+    //    <release>1.13.0</release>
+    //  </versioning>
+    //</metadata>
+    Element metadataElement = document.getDocumentElement();
+    Element versioningElement = getFirstChildElement(metadataElement, "versioning");
+    if (versioningElement != null) {
+      releaseVersion = versioningElement.getElementsByTagName("release").item(0).getTextContent();
+    } else {
+      verboseLog(String.format("Could not find <versioning> tag for %s, returning null version", url));
+    }
+    return releaseVersion;
+  }
+
+  public static Element getFirstChildElement(Element element, String tagName) {
+    NodeList nodeList = element.getElementsByTagName(tagName);
+    for (int i = 0; i < nodeList.getLength(); i++)
+    {
+      Node node = nodeList.item(i);
+      if (node.getNodeType() == Node.ELEMENT_NODE)
+      {
+        return (Element) node;
+      }
+    }
+    return null;
+  }
+
+  public static void verboseLog(String logline) {
+    if (System.getenv("RJE_VERBOSE") != null) {
+      System.out.println(logline);
+    }
+  }
+
+  public static void main(String[] args) throws IOException {
+    verboseLog(String.format("Running outdated with args %s", Arrays.toString(args)));
+
+    if (args.length != 2) {
+      System.out.println("Usage: outdated <artifact_file_path> <repositories_file_path");
+      System.exit(1);
+    }
+
+    String artifactsFilePath = args[0];
+    String repositoriesFilePath = args[1];
+
+    List<String> artifacts = Files.readAllLines(Paths.get(artifactsFilePath), StandardCharsets.UTF_8);
+    List<String> repositories = Files.readAllLines(Paths.get(repositoriesFilePath), StandardCharsets.UTF_8);
+
+    System.out.println(String.format("Checking for updates of %d artifacts against the following repositories:", artifacts.size()));
+    for (String repository: repositories) {
+      System.out.println(String.format("\t%s", repository));
+    }
+    System.out.println();
+
+    boolean foundUpdates = false;
+
+    for (String artifact: artifacts) {
```

Comment:
It should also be straightforward to make this multithreaded if we wanted to check for several updates at once and reduce runtime.

Reply:
Add this note as a comment?